### ISSUE TITLE
Why Precision@0.50 is not at all improving? ideas to improve it?

### ISSUE THREAD

**Issue by isri-ram** [Deci-AI/super-gradients#1734]
### 💡 Your Question

I'm training a yolo-nas-s model for object detection with 14 classes.
Train - 3076 images
Test - 879 images
Valid - 439 images
Total - 4393 images
Earlier i trained a medium as well as a large model. Even after 100 epochs the Precision@0.50 is below 0.10. So now i'm trying the small version of yolo nas. 
I have applied transforms as well (if that's necessary to be known)
```
train_data.dataset.transforms
```
Below is my train params:
```
from super_gradients.training.losses import PPYoloELoss
from super_gradients.training.metrics import DetectionMetrics_050
from super_gradients.training.models.detection_models.pp_yolo_e import PPYoloEPostPredictionCallback

train_params = {
    "average_best_models":True,
    "warmup_mode": "LinearBatchLRWarmup",
    "warmup_initial_lr": 1e-6,
    "lr_warmup_steps":1000,
    "lr_warmup_epochs": 3,
    "initial_lr": 2e-4,
    "lr_mode": "CosineLRScheduler",
    "cosine_final_lr_ratio": 0.1,
    "optimizer": "AdamW",
    "optimizer_params": {"weight_decay": 0.00001},
    "zero_weight_decay_on_bias_and_bn": True,
    "batch_accumulate":1,
    "ema": True,
    "ema_params": {"decay": 0.997, "decay_type": "threshold"},
    "max_epochs": 500,
    "mixed_precision": True,
    "loss": PPYoloELoss(
        use_static_assigner=False,
        num_classes=len(dataset_params['classes'])
    ),
    "valid_metrics_list": [
        DetectionMetrics_050(
            score_thres=0.1,
            top_k_predictions=300,
            num_cls=len(dataset_params['classes']),
            normalize_targets=True,
            post_prediction_callback=PPYoloEPostPredictionCallback(
                score_threshold=0.01,
                nms_top_k=1000,
                max_predictions=300,
                nms_threshold=0.7,
            )
        )
    ],
    "metric_to_watch": 'mAP@0.50',
    "greater_metric_to_watch_is_better": True
}
```
This below metric is at 50th epoch. The recall is better than the precision. **I know this is due to false positives.**
```
SUMMARY OF EPOCH 50
├── Train
│   ├── Ppyoloeloss/loss_cls = 1.1821
│   │   ├── Epoch N-1      = 1.1859 (↘ -0.0038)
│   │   └── Best until now = 1.1859 (↘ -0.0038)
│   ├── Ppyoloeloss/loss_iou = 0.6381
│   │   ├── Epoch N-1      = 0.6395 (↘ -0.0015)
│   │   └── Best until now = 0.6395 (↘ -0.0015)
│   ├── Ppyoloeloss/loss_dfl = 0.6101
│   │   ├── Epoch N-1      = 0.6108 (↘ -0.0007)
│   │   └── Best until now = 0.6108 (↘ -0.0007)
│   └── Ppyoloeloss/loss = 2.4303
│       ├── Epoch N-1      = 2.4363 (↘ -0.006)
│       └── Best until now = 2.4363 (↘ -0.006)
└── Validation
    ├── Ppyoloeloss/loss_cls = 1.1855
    │   ├── Epoch N-1      = 1.1889 (↘ -0.0034)
    │   └── Best until now = 1.1816 (↗ 0.0039)
    ├── Ppyoloeloss/loss_iou = 0.5816
    │   ├── Epoch N-1      = 0.5829 (↘ -0.0013)
    │   └── Best until now = 0.5829 (↘ -0.0013)
    ├── Ppyoloeloss/loss_dfl = 0.5778
    │   ├── Epoch N-1      = 0.5779 (↘ -1e-04)
    │   └── Best until now = 0.5779 (↘ -1e-04)
    ├── Ppyoloeloss/loss = 2.3449
    │   ├── Epoch N-1      = 2.3496 (↘ -0.0048)
    │   └── Best until now = 2.3496 (↘ -0.0048)
    ├── Precision@0.50 = 0.034
    │   ├── Epoch N-1      = 0.0352 (↘ -0.0012)
    │   └── Best until now = 0.042  (↘ -0.008)
    ├── Recall@0.50 = 0.606
    │   ├── Epoch N-1      = 0.5977 (↗ 0.0083)
    │   └── Best until now = 0.6036 (↗ 0.0024)
    ├── Map@0.50 = 0.2356
    │   ├── Epoch N-1      = 0.2337 (↗ 0.0019)
    │   └── Best until now = 0.2354 (↗ 0.0002)
    └── F1@0.50 = 0.063
        ├── Epoch N-1      = 0.0627 (↗ 0.0004)
        └── Best until now = 0.0653 (↘ -0.0023)
```
**Any ideas on how to improve precision?**

**Comment by isri-ram** [comment]
Hey @BloodAxe any thoughts to improve Precision@0.50 ?

**Comment by BloodAxe** [comment]
We evaluate top 300 detections per image. In COCOEval or in YoloV8 the top 20 detections are checked. The 300 detections boost mAP score a bit. Unfortunately a precision is affected.
So what you can do here:

1) Decrease number of detections to evaluate: 
```
DetectionMetrics:
  top_k_predictions: 20
  ...
```

2) Enable threshold optimization - this will search the best confidence threshold that maximize the F1 score. 

```
DetectionMetrics:
  calc_best_score_thresholds: True
  ...
```

**Comment by isri-ram** [comment]
Hey @BloodAxe @ShpihanVlad thanks for your inputs. I experimented with whatever parameters you said. 

I also experimented with `DetectionMetrics_050` and `PPYoloEPostPredictionCallback`. This did help improve but if precision increases then recall decreases and when recall increases the precision decreases. I couldn't strike a perfect balance between them.

Also the `Ppyoloeloss` is high. Specifically the `Ppyoloeloss/loss_cls` is higher than other losses.

I'm using Roboflow's VinBigdata Chest X-ray images for anomaly detection (object detection) with 14 classes.
Train - 3076 images
Test - 879 images
Valid - 439 images
Total - 4393 images

My Train and Validation parameters:
```
train_params = {
    "average_best_models":True,
    "warmup_mode": "LinearBatchLRWarmup",
    "warmup_initial_lr": 1e-5,
    "lr_warmup_steps":192,
    "lr_warmup_epochs": 3,
    "initial_lr": 1e-3,
    "lr_mode": "CosineLRScheduler",
    "cosine_final_lr_ratio": 0.1,
    "optimizer": "Adam",
    "optimizer_params": {"weight_decay": 0.00001},
    "zero_weight_decay_on_bias_and_bn": True,
    "batch_accumulate":1,
    "ema": True,
    "ema_params": {"decay": 0.997, "decay_type": "threshold"},
    "max_epochs": 500,
    "mixed_precision": True,
    "loss": PPYoloELoss(
        use_static_assigner=False,
        num_classes=len(dataset_params['classes'])
    ),
    "valid_metrics_list": [
        DetectionMetrics_050(
            score_thres=0.4,
            top_k_predictions=20,
            calc_best_score_thresholds= True,
            num_cls=len(dataset_params['classes']),
            normalize_targets=True,
            post_prediction_callback=PPYoloEPostPredictionCallback(
                score_threshold=0.4,
                nms_top_k=140,
                max_predictions=200,
                nms_threshold=0.6,
            )
        )
    ],
    "metric_to_watch": 'mAP@0.50',
    "greater_metric_to_watch_is_better": True
}
```
Below is the summary of my model performance at epoch 18:
```
SUMMARY OF EPOCH 18
├── Train
│   ├── Ppyoloeloss/loss_cls = 1.2225
│   │   ├── Epoch N-1      = 1.2202 (↗ 0.0023)
│   │   └── Best until now = 1.2202 (↗ 0.0023)
│   ├── Ppyoloeloss/loss_iou = 0.6844
│   │   ├── Epoch N-1      = 0.6879 (↘ -0.0035)
│   │   └── Best until now = 0.6876 (↘ -0.0032)
│   ├── Ppyoloeloss/loss_dfl = 0.686
│   │   ├── Epoch N-1      = 0.6885 (↘ -0.0025)
│   │   └── Best until now = 0.6883 (↘ -0.0023)
│   └── Ppyoloeloss/loss = 2.593
│       ├── Epoch N-1      = 2.5966 (↘ -0.0036)
│       └── Best until now = 2.5966 (↘ -0.0036)
└── Validation
    ├── Ppyoloeloss/loss_cls = 1.2152
    │   ├── Epoch N-1      = 1.2138 (↗ 0.0014)
    │   └── Best until now = 1.2111 (↗ 0.0041)
    ├── Ppyoloeloss/loss_iou = 0.612
    │   ├── Epoch N-1      = 0.6158 (↘ -0.0037)
    │   └── Best until now = 0.6158 (↘ -0.0037)
    ├── Ppyoloeloss/loss_dfl = 0.6414
    │   ├── Epoch N-1      = 0.6467 (↘ -0.0053)
    │   └── Best until now = 0.6448 (↘ -0.0033)
    ├── Ppyoloeloss/loss = 2.4687
    │   ├── Epoch N-1      = 2.4763 (↘ -0.0076)
    │   └── Best until now = 2.4719 (↘ -0.0032)
    ├── Precision@0.50 = 0.1792
    │   ├── Epoch N-1      = 0.1784 (↗ 0.0008)
    │   └── Best until now = 0.1784 (↗ 0.0008)
    ├── Recall@0.50 = 0.1255
    │   ├── Epoch N-1      = 0.1254 (↗ 0.0001)
    │   └── Best until now = 0.1287 (↘ -0.0032)
    ├── Map@0.50 = 0.1169
    │   ├── Epoch N-1      = 0.1169 (↘ -0.0)
    │   └── Best until now = 0.1172 (↘ -0.0003)
    ├── F1@0.50 = 0.1163
    │   ├── Epoch N-1      = 0.1159 (↗ 0.0004)
    │   └── Best until now = 0.1159 (↗ 0.0004)
    ├── Best_score_threshold = 0.0
    │   ├── Epoch N-1      = 0.41   (↘ -0.41)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_0 = 0.52
    │   ├── Epoch N-1      = 0.52   (= 0.0)
    │   └── Best until now = 0.51   (↗ 0.01)
    ├── Best_score_threshold_cls_1 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_2 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_3 = 0.0
    │   ├── Epoch N-1      = 0.42   (↘ -0.42)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_4 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_5 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_6 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_7 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_8 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_9 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_10 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_11 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    ├── Best_score_threshold_cls_12 = 0.0
    │   ├── Epoch N-1      = 0.0    (= 0.0)
    │   └── Best until now = 0.0    (= 0.0)
    └── Best_score_threshold_cls_13 = 0.0
        ├── Epoch N-1      = 0.0    (= 0.0)
        └── Best until now = 0.0    (= 0.0)
```

How do I strike a balance between Precision and Recall?
How do I reduce `Ppyoloeloss` ?

**Comment by BloodAxe** [comment]
I see a number of troubling issues in the configuration file you've shared:

```
    "valid_metrics_list": [
        DetectionMetrics_050(
            score_thres=0.4,
            top_k_predictions=20,
            calc_best_score_thresholds= True,
            num_cls=len(dataset_params['classes']),
            normalize_targets=True,
            post_prediction_callback=PPYoloEPostPredictionCallback(
                score_threshold=0.4,
                nms_top_k=140,
                max_predictions=200,
                nms_threshold=0.6,
            )
        )
    ],
```

* Number of predictions before NMS is lower than the total number of predictions. It does not make any sense, as you want the number of `nms_top_k` to be greater than `max_predictions` since some predictions will be removed by NMS step. So you may set `nms_top_k` to 256 or 512 and `max_predictions` to 100 or even 20.
* The `score_threshold=0.4` sets the quite high bar and filter out any prediction that has score lower than 0.4 and this filtering happens as a first step in candidate selection for NMS. So in `post_prediction_callback` you usually want to set it to something low (0.05). 
* Since you have `calc_best_score_thresholds=True`, you should set `score_thres` to the same value as `score_threshold` to use all available predictions (Those with scores > `score_threhsold`) to optimize mean F1 score.
* Tip - you can specify `class_names` array to give meaningful names to classes in the reported metric if you want.


The is another issue I can spot which is zero score for all but one class in your training. This looks extremely suspicious and you should double-check whether dataset you've downloaded is indeed correct. 
The [Kaggle version](https://www.kaggle.com/c/vinbigdata-chest-xray-abnormalities-detection/code) of this dataset has 15K images in total so the total number you report (4393) is barely a 1/3 and I don't know how Roboflow made their version. Maybe some classes are just missing in their version or maybe there is a bug. I suggest to double-check the classes presence in train/val splits that you are using and ensure all classes are presented in each split.
